src/system_info/system_info_network_ivi.cc

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2014 Intel Corporation. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "system_info/system_info_network.h"
+
+SysInfoNetwork::SysInfoNetwork()
+    : type_(SYSTEM_INFO_NETWORK_UNKNOWN) {
+  PlatformInitialize();
+}
+
+void SysInfoNetwork::PlatformInitialize() {
+  conn_ = system_info::GetDbusConnection();
+  if (!conn_)
+    return;
+  modem_path_ = system_info::OfonoGetModemPath(conn_);
+}
+
+SysInfoNetwork::~SysInfoNetwork() {
+  if (!conn_)
+    return;
+  g_dbus_connection_signal_unsubscribe(conn_,
+      cellular_network_type_changed_watch_);
+  g_dbus_connection_close_sync(conn_, NULL, NULL);
+  conn_ = NULL;
+}
+
+void SysInfoNetwork::StartListening() {
+  vconf_notify_key_changed(VCONFKEY_NETWORK_STATUS,
+      (vconf_callback_fn)OnNetworkTypeChanged, this);
+
+  if (!conn_ || !modem_path_)
+    return;
+  cellular_network_type_changed_watch_ = g_dbus_connection_signal_subscribe(
+      conn_, system_info::kOfonoService,
+      system_info::kOfonoNetworkRegistrationIface, "PropertyChanged",
+      modem_path_, NULL, G_DBUS_SIGNAL_FLAGS_NONE,
+      OnCellularNetworkTypeChanged, this, NULL);
+}
+
+void SysInfoNetwork::StopListening() {
+  vconf_ignore_key_changed(VCONFKEY_NETWORK_STATUS,
+      (vconf_callback_fn)OnNetworkTypeChanged);
+
+  if (!conn_)
+    return;
+  g_dbus_connection_signal_unsubscribe(conn_,
+      cellular_network_type_changed_watch_);
+}
+
+bool SysInfoNetwork::Update(picojson::value& error) {
+  if (GetNetworkType())
+    return true;
+
+  system_info::SetPicoJsonObjectValue(error, "message",
+      picojson::value("Get network type faild."));
+  return false;
+}
+
+bool SysInfoNetwork::GetNetworkType() {
+  int status = 0;
+  if (vconf_get_int(VCONFKEY_NETWORK_STATUS, &status))
+    return false;
+
+  switch (status) {
+    case VCONFKEY_NETWORK_WIFI:
+      type_ = SYSTEM_INFO_NETWORK_WIFI;
+      break;
+    case VCONFKEY_NETWORK_ETHERNET:
+      type_ = SYSTEM_INFO_NETWORK_ETHERNET;
+      break;
+    case VCONFKEY_NETWORK_BLUETOOTH:
+      type_ = SYSTEM_INFO_NETWORK_UNKNOWN;
+      break;
+    case VCONFKEY_NETWORK_CELLULAR:
+      if (!GetCellularNetworkType())
```

Comment:
It seems useless define <code>GetCellularNetworkType</code>'s return value as <code>bool</code>. The only use is here. But you can do <code>type_ = SYSTEM_INFO_NETWORK_UNKNOWN;</code>inside  <code>GetCellularNetworkType</code>